common/netlink.cpp

```diff
@@ -0,0 +1,468 @@
+        AttributeList attributes = linkinfo.second;
+        for (AttributeInfo attrinfo : attributes) {
+            rtattr attr = attrinfo.first;
+            void *data = attrinfo.second;
+
+            if (attr.rta_type == IFLA_IFNAME) {
+                char *ifname = (char *) data;
+                if (strcmp(ifname, bridgeName) == 0) {
+                    hasBridge = true;
+                    bridge_ifindex = linkinfo.first.ifi_index;
+                    break;
+                }
+            }
+        }
+
+        // TODO: Make this nicer, please
+        if (hasBridge) {
+            break;
+        }
+    }
+
+    if (!hasBridge) {
+        return false;
+    }
+
+	bool bridgeHasGateway = false;
+    for (AddressInfo addressInfo : m_addresses) {
+        ifaddrmsg addrmsg = addressInfo.first;
+        if (addrmsg.ifa_index != bridge_ifindex) {
+            continue;
+        }
+
+        AttributeList attributes = addressInfo.second;
+        for (AttributeInfo attrPair : attributes) {
+            rtattr attr = attrPair.first;
+            if (attr.rta_type != IFA_ADDRESS && attr.rta_type != IFA_LOCAL) {
+                continue;
+            }
+
+            void *data = attrPair.second;
+            char out[INET6_ADDRSTRLEN];
+            // It is ok here for inet_ntop to fail (data could be bad)
+            if (inet_ntop(addrmsg.ifa_family, data, out, sizeof(out))) {
+                if (strcmp(out, expectedAddress) == 0) {
+                    bridgeHasGateway = true;
+                    break;
+                }
+            }
+        }
+
+        if (bridgeHasGateway) {
+            break;
+        }
+    }
+
+    return bridgeHasGateway;
+}
+
+int Netlink::read_msg()
+{
+    bool end = false;
+	size_t PAGE_SIZE = getpagesize();
+    char buf[PAGE_SIZE];
+
+    while (!end) {
+		int len;
+		struct nlmsghdr *msg;
+		struct msghdr reply;
+		memset(&reply, 0, sizeof(reply));
+
+        struct iovec io = { buf, PAGE_SIZE };
+		reply.msg_iov = &io;
+		reply.msg_iovlen = 1;
+		reply.msg_name = &m_kernel;
+		reply.msg_namelen = sizeof(m_kernel);
+
+		len = recvmsg(m_fd, &reply, 0); /* read lots of data */
+		if (len > 0) {
+			for (msg = (struct nlmsghdr *) buf; NLMSG_OK(msg, len); msg = NLMSG_NEXT(msg, len)) {
+				switch(msg->nlmsg_type)
+				{
+                    case NLMSG_ERROR: {
+                        struct nlmsgerr *err = (struct nlmsgerr *)NLMSG_DATA(msg);
+                        if (err->error != 0) {
+                            return err->error; // Actually an error
+                        } else {
+                            end = true; // This was an ack
+                        }
+
+                        break;
+                    }
+					case NLMSG_DONE:
+						end = true;
+						break;
+                    case RTM_GETLINK:
+					case RTM_NEWLINK:
+                        save_msg<ifinfomsg, LinkInfo>(msg, m_links);
+                        break;
+                    case RTM_NEWADDR:
```

Comment:
All the breaks feel odd, gather them all up at the end with a default handler perhaps?